### PR TITLE
Add tests for env reqs

### PR DESCRIPTION
**Pull request by segiddins**
\c @indirect

### CODE REVIEW

**Comment by segiddins** [comment]
I don't know why this is only failing on 1.8.7 :/

File: test/rubygems/test_gem.rb [hunk 1]
```diff
@@ -777,6 +777,22 @@ def test_self_ruby_api_version
     RbConfig::CONFIG['ruby_version'] = orig_ruby_version
   end
 
+  def test_self_env_requirement
+    old_env = ENV.to_hash
+
+    ENV.clear
```
Comment (drbrain) [inline test/rubygems/test_gem.rb]: RubyGems tests are supposed to run with `$SAFE = 1` but only ruby 1.8.7 does this.  You cannot wipe ENV with `$SAFE = 1` so you should instead clear these in setup and teardown (missing) [up here](https://github.com/segiddins/rubygems/blob/seg-env-req-tests/test/rubygems/test_gem.rb#L19).

**Comment by segiddins** [comment]
Ought to be fixed now

**Comment by mpapis** [comment]
I just wonder if this is not another way of defining dependencies when we already have 3 formats of gem dependency files supported in rubygems? maybe another fake format (that loads the variables) would be better maybe `RUBYGEMS_GEMDEPS="rake>=10:bundler=` handled here: https://github.com/rubygems/rubygems/blob/master/lib/rubygems.rb#L1077-L1084

**Comment by mpapis** [comment]
`SecurityError: Insecure operation` this happens only on Travis Ruby 1.8.7 because of `$SAFE=1` in https://github.com/rubygems/rubygems/blob/master/test/rubygems/test_gem.rb#L10 - check if commenting it out would solve the problem, maybe `$SAFE=1 unless RUBY_VERSION == "1.8.7"`?

**Comment by segiddins** [comment]
@mpapis this is only adding tests, the feature was already merged

@drbrain you said moving the unsafe operations to `setup` / `teardown` would make this pass...?

**Comment by tenderlove** [comment]
```diff
commit e08735df7221a384d33fc85f076058445a514d28
Author: Aaron Patterson <aaron.patterson@gmail.com>
Date:   Tue Mar 1 11:21:52 2016 -0800

    Fix 1.8 support in the tests

diff --git a/lib/rubygems/test_case.rb b/lib/rubygems/test_case.rb
index 5a0aa1d..e825533 100644
--- a/lib/rubygems/test_case.rb
+++ b/lib/rubygems/test_case.rb
@@ -223,7 +223,10 @@ class Gem::TestCase < MiniTest::Unit::TestCase
     @orig_gem_spec_cache = ENV['GEM_SPEC_CACHE']
     @orig_rubygems_gemdeps = ENV['RUBYGEMS_GEMDEPS']
     @orig_rubygems_host = ENV['RUBYGEMS_HOST']
-    @orig_gem_env_requirements = ENV.delete_if { |k, _v| k.start_with?('GEM_REQUIREMENT_') }
+    ENV.keys.find_all { |k| k.start_with?('GEM_REQUIREMENT_') }.each do |k|
+      ENV.delete k
+    end
+    @orig_gem_env_requirements = ENV.to_hash
 
     ENV['GEM_VENDOR'] = nil
 
@@ -370,6 +373,11 @@ class Gem::TestCase < MiniTest::Unit::TestCase
 
     FileUtils.rm_rf @tempdir unless ENV['KEEP_FILES']
 
+    ENV.clear
+    @orig_gem_env_requirements.each do |k,v|
+      ENV[k] = v
+    end
+
     ENV['GEM_HOME']   = @orig_gem_home
     ENV['GEM_PATH']   = @orig_gem_path
     ENV['GEM_VENDOR'] = @orig_gem_vendor
@@ -377,9 +385,6 @@ class Gem::TestCase < MiniTest::Unit::TestCase
     ENV['RUBYGEMS_GEMDEPS'] = @orig_rubygems_gemdeps
     ENV['RUBYGEMS_HOST'] = @orig_rubygems_host
 
-    ENV.delete_if { |k, _v| k.start_with?('GEM_REQUIREMENT_') }
-    ENV.update(@orig_gem_env_requirements)
-
     Gem.ruby = @orig_ruby if @orig_ruby
 
     if @orig_ENV_HOME then
```

This fixes it for me.  I'm not sure how (or if I can) push to this PR, but you can find the commit [here](https://github.com/tenderlove/rubygems/commit/e08735df7221a384d33fc85f076058445a514d28).

**Comment by indirect** [comment]
@segiddins can you cherry pick the fix into this PR? I'm 👍 once this passes

**Comment by segiddins** [comment]
Ought to be fixed now. Feel free to r on green

**Comment by indirect** [comment]
@segiddins awesome, thanks!

@homu r+

**Pull request merged**